feat: Consensus-based slashing proposer contract

spalladino opened the pull request:
Implementation for the consensus-based slashing on L1 as discussed [here](https://linear.app/aztec-labs/document/slashing-discussion-w33-8810b51b11b0#l1-consensus-db4c0919):

> Build a new SlashingProposer version that does not extend from EmpireBase, but follows same logic on slashing rounds, execution delays, and allowing votes from the current L2 proposer only.

> Each vote is now an array of proposed slashes. We define a slashing-unit value, and every slash is a multiple of this unit. Slashes are represented as a bit vector, where the i-th 4-bits (or 2, or 8, depending on how much granularity we want) indicate how many slashing-units proposer i is slashed. An expensive tallying function computes the effective payload by iterating over all votes.

This still keeps the EmpireSlashProposer around. Next step will be to let the `Slasher` choose which implementation to use, so we can keep tests for both, and then rebase this on top of #16284 and tweak the updated slasher client to use this new SlashProposer.

Other changes:
- Updates the `SlashProposer` to be `Ownable` by governance, who can execute slash payloads.
- Modifies the `Slasher` deployment due to contract size limits (the slasher no longer deploys the proposer, instead it needs to be initialized after deployment with the proposer address).
- Renames `ExtRollupLibN` to have more descriptive names.
- Adds env var config for the new proposer flavor.
- Adds type-checking for L1 contract deployer on ts.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+    Slot lastVoteSlot; // Last slot for which a vote was received
+    bool executed; // Whether this round has been executed
+    // QUESTION: Each element of this array is defined as a dynamically-sized `bytes` array, but we know its size is
+    // `COMMITTEE_SIZE * ROUND_SIZE_IN_EPOCHS / 2` exactly. Can we use that to optimize how this is stored?
```
Could we not make it fixed size [][] so it is not necessary to keep track of the size.

spalladino replied:
Stupid question: how do you make a fixed size `bytes` array? AFAIU the padding in `byte1[SIZE]` is awful.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+   * @param _epochIndex The index of the epoch within the round (0 to ROUND_SIZE_IN_EPOCHS-1)
+   */
+  function getCommitteeCommitment(Round _round, uint256 _epochIndex) internal returns (bytes32) {
+    // TODO: Check if there is a less convoluted way to get the timestamp to get the committee commitment
```
There is `toTimestamp` on the epoch, so would think you could just do?

```solidity
Timestamp timestamp = Epoch.wrap(_round * ROUND_SIZE_IN_EPOCHS).toTimestamp();
```

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+    );
+
+    // Store the vote for this round and update round data
+    // QUESTION: Does this trigger multiple separate storage writes? Can we merge them somehow?
```
Right now it would because of the sizes. If you make them samller though, and keep track of the variable size input in a different map you could probably do 
```solidity
roundData = RoundData({});
```
Would help you on the `lastVoteSlot´ and `voteCount`, the `votes` still sepate.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+
+    // Get the committee commitment from the Rollup at this timestamp
+    IValidatorSelection rollup = IValidatorSelection(INSTANCE);
+    (bytes32 commitment,) = rollup.getCommitteeCommitmentAt(timestamp);
```
Just a note, but potentially samples

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+   */
+  function getRound(Round _round) external view returns (bool isExecuted, bool readyToExecute, uint256 voteCount) {
+    // Load round data from the circular storage
+    RoundData storage roundData = getRoundData(_round);
```
Might be stale

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+    RoundData storage roundData = getRoundData(_round);
+
+    // If the round data is stale (not written for this round), return empty array
+    if (Round.unwrap(roundData.roundNumber) != Round.unwrap(_round)) {
```
Why would we not just revert it? e.g., use `loadRoundDate` instead?

spalladino replied:
Because this function was originally meant to be tagged as `view`, but I had to remove it when adding the `getEpochCommittee` call. So yeah, I'll switch to loadRoundData. Good catch.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+    if (Round.unwrap(roundData.roundNumber) != Round.unwrap(_round)) {
+      // Reset the round data for this round. Note that we do NOT delete the votes from the array, since they are
+      // implicitly reset by setting vote count to zero.
+      // QUESTION: Does this trigger multiple separate storage writes? Can we merge them?
```
Atm yes because storage size, but as earlier, if moving the votes out you can more easily set it at once. One of those where using a `type` can be useful for compression if the struct is annoying.

LHerskind reviewed (commented):
Mainly looked at the questions haven't looked deep at all, entirely skipping most of the files entirely and most of the content in `ConsensusSlashingProposer`

just-mitch commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,560 @@
+    }
+
+    // Create properly sized array with only the actual slash actions
+    // QUESTION: Do we actually need to copy this? Cant we do some assembly magic? Slices dont seem to work here.
```
I accidentally deleted a comment from @LHerskind. He points out that you can
```
assembly {
  mstore(actions, actionCount)
}
```
to adjust the length of `actions` and then just return `actions`.

LHerskind commented on l1-contracts/src/core/libraries/rollup/ConsensusDeploymentExtLib.sol:
```diff
@@ -0,0 +1,54 @@
+import {ConsensusSlashingProposer} from "@aztec/core/slashing/ConsensusSlashingProposer.sol";
+
+/**
+ * @title ExtRollupLib4 - External Rollup Library (Consensus Slasher Deployment)
```
Collecting External Rollup Libraries like Thanos collecting infinity stones.

![Marvel Balance GIF](https://media2.giphy.com/media/Ry1MOAeAYXvRVQLPw3/giphy.gif)

LHerskind commented on l1-contracts/src/core/RollupCore.sol:
```diff
@@ -228,21 +231,42 @@ contract RollupCore is EIP712("Aztec Rollup", "1"), Ownable, IStakingCore, IVali
+    // Deploy slasher based on flavor
+    ISlasher slasher;
+
+    if (_config.slasherFlavor == SlasherFlavor.CONSENSUS) {
```
Did the if here not fit inside the external lib? 👀

spalladino replied:
Thing is I needed to call different external libs depending on the flavor, since each lib handles a different deployment.

LHerskind replied:
Oh, ye, but were more, why have multiple external libs, you could have one lib that deploys any of them, would not expect that to be over the limits.

spalladino replied:
> you could have one lib that deploys any of them, would not expect that to be over the limits.

Surprisignly enough, it did go over the limits. I originally had only one lib that deployed them all, and it exceeded contract size limit.

LHerskind commented on l1-contracts/src/shared/libraries/TimeMath.sol:
```diff
@@ -8,6 +8,8 @@ type Slot is uint256;
 
 type Epoch is uint256;
 
+type SlashRound is uint256;
```
I would prefer if we do not put the `SlashRound` in here as it is not really shared but only used in one of the settings, and because it is extending already reviewed contracts without really needing to be in there.

I acknowledge that really the Epoch should probably not have been in here either.

spalladino replied:
Moved to `src/shared/libraries/SlashRoundLib.sol`

LHerskind replied:
Know I am a pain here, and the slashing fits kinda in that fuzzy space between gov and core, but should we just have this in core as the tally itself is in core and nothing else seems to use it?

spalladino replied:
Moved to `src/core/libraries/SlashRoundLib.sol`

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -222,16 +241,20 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
 
-  function test_100_slashing_validators() public prepare(100, false) {
-    benchmark(true);
+  function test_100_empire_slashing_validators() public prepare(100, false, TestSlash.EMPIRE) {
```
If we want to run these separate, we should update the .py that we use to generate the descriptions as well. 

See the `generate_benchmark_json.py` and `gas_benchmark.py`, a hurdle would be that both end up producing gas reports for `aggregate3` so would likely need some more playing around.

I think it is fine to not necessarily focus too much on that now though, as it is not a thing I want to block this one 😬 and it seems like something where updating a bit earlier as part of the `gas_benchmark.py` input could make it fairly simple (but some maybe annoying updates in there that I can deal with later)

spalladino replied:
Ngl I went through those scripts and didn't understand much. To avoid sinking too much time in them, I'll remove the empire slashing benchmark (since we won't be using it) and just run the consensus one.

LHerskind replied:
No problem 👍

LHerskind commented on l1-contracts/src/shared/libraries/CompressedTimeMath.sol:
```diff
@@ -3,14 +3,16 @@
 
 import {SafeCast} from "@oz/utils/math/SafeCast.sol";
-import {Timestamp, Slot, Epoch} from "./TimeMath.sol";
+import {Timestamp, Slot, Epoch, SlashRound} from "./TimeMath.sol";
```
Same as https://github.com/AztecProtocol/aztec-packages/pull/16357/files#r2286188799

spalladino replied:
Moved to the same `SlashRoundLib` library

LHerskind commented on l1-contracts/src/core/slashing/EmpireSlashingProposer.sol:
```diff
@@ -10,16 +10,22 @@ import {EmpireBase} from "@aztec/governance/proposer/EmpireBase.sol";
+  /**
+   * @notice Type of slashing proposer (either Consensus or Empire)
+   */
+  string public constant SLASHING_PROPOSER_TYPE = "Empire";
```
Why not use the flavor enum?

spalladino replied:
I wanted to get something more explicit than 0 or 1 when calling from off-chain. But I can switch to the enum if we think it's cleaner.

LHerskind commented on l1-contracts/src/core/slashing/Slasher.sol:
```diff
@@ -3,45 +3,46 @@
+  // solhint-disable-next-line comprehensive-interface
+  function initializeProposer(address _proposer) external {
+    require(PROPOSER == address(0), Slasher__AlreadyInitialized());
+    require(_proposer != address(0), Slasher__ProposerZeroAddress());
```
Would usually flip the ordering to avoid the `sload` that might be unnecessary when `_proposer = address(0)`, but as this is called once, it is a micro nit really.

spalladino replied:
And besides, if you call without checking preconditions, it's your fault - you're lucky I'm doing a require and not an assert to burn all your gas

LHerskind commented on l1-contracts/src/core/slashing/Slasher.sol:
```diff
@@ -3,45 +3,46 @@
 
-contract Slasher is ISlasher {
-  SlashingProposer public immutable PROPOSER;
+contract Slasher is ISlasher, Ownable {
```
With the introduction of the owner here, and it only being used for the slash as I see it here, should it allow also updating the `vetoer`? The gov could do that anyway by doing a new slasher, but if we are going on the path of making everything directly possible by gov we could alter that in here 🤷

spalladino replied:
I originally added a GOVERNANCE immutable, but switched to Ownable when I saw it was the pattern used in the Rollup. I'm fine with either.

spalladino replied:
Removed Ownable in favor of GOVERNANCE immutable

LHerskind commented on l1-contracts/src/core/slashing/Slasher.sol:
```diff
@@ -3,45 +3,46 @@
   function vetoPayload(IPayload _payload) external override(ISlasher) returns (bool) {
     require(msg.sender == VETOER, Slasher__CallerNotVetoer(msg.sender, VETOER));
     vetoedPayloads[address(_payload)] = true;
+    emit VetoedPayload(address(_payload));
```
Thanks.

spalladino replied:
Still got to update the slasher-client so it does not try to execute a payload that has been vetoed. There's nothing on the node that monitors these events or checks the vetoed payloads.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+  /**
+   * @notice Type of slashing proposer (either Consensus or Empire)
+   */
+  string public constant SLASHING_PROPOSER_TYPE = "Consensus";
```
Why not the enum?

spalladino replied:
Switched to enum

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   * @notice How many rounds in the past to look when determining which validators to slash
+   * @dev Creates temporal separation between the time of misbehavior and voting for slashing
+   */
+  uint256 public immutable SLASH_OFFSET_IN_ROUNDS;
```
Essentially just a value we use to convey information to the nodes, ye?

spalladino replied:
It's used when determining the validators from which epochs get slashed in round N

spalladino replied:
Clarified in a comment

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   *      and then adding the epoch index within that round.
+   *
+   * @param _round The round number during which voting is taking place
+   * @param _epochIndex The index of the epoch within the round (must be 0 to ROUND_SIZE_IN_EPOCHS-1)
```
Why not require this if it is a must? 

```solidity
require(_epochIndex < ROUND_SIZE_IN_EPOCHS);`
```

spalladino replied:
Added

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   * @param round The round number in which the vote was cast
+   * @param proposer The address of the proposer who cast the vote
+   */
+  event VoteCast(SlashRound indexed round, address indexed proposer);
```
Should anything related to the vote also be emitted?

spalladino replied:
I could emit the slot number, but there is no useful info on the vote itself to be emitted. Maybe total slashes? A hash of the votes? Nothing useful to read from the node tbh.

spalladino replied:
Added slot number

LHerskind replied:
> Nothing useful to read from the node tbh.

Not necessarily for the node, but also observers and analysis etc 🤷

spalladino replied:
Got it. Still, I think slot number is pretty much the only other useful info we could get.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+    bytes32 digest = getVoteSignatureDigest(_votes, slot);
+    require(_sig.verify(proposer, digest), Errors.ConsensusSlashingProposer__InvalidSignature());
+
+    // Each byte encodes 2 validators (4 bits each), so we need half the total validators in bytes
```
So each validator is a nibble

spalladino replied:
Yep

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+ *      - SLASHING_UNIT: Base amount of slashing per unit voted
+ *      - COMMITTEE_SIZE: Number of validators per committee
+ */
+contract ConsensusSlashingProposer is EIP712 {
```
I don't really get the `Consensus` naming, so was earlier things 🤷

spalladino replied:
Not a fan either, but it's now everywhere as I couldn't think of a better idea. The idea is that validators come to consensus on who to slash via this contract.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+        LIFETIME_IN_ROUNDS, EXECUTION_DELAY_IN_ROUNDS
+      )
+    );
+    require(
```
🫡

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+    // Store the vote for this round and update round data
+    uint256 voteCount = roundData.voteCount;
+    getRoundVotes(round).votes[voteCount] = _votes;
+    setRoundData(round, slot, voteCount + 1, roundData.executed);
```
The `roundData.executed` looks pretty weird just as it almost feels like it will be updated, but we would never vote if already executed etc.

spalladino replied:
Clarified in a comment

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   * Reverts with:
+   * - ConsensusSlashingProposer__VotingNotOpen: If the round is less than SLASH_OFFSET_IN_ROUNDS
+   */
+  function getEpochSlashed(SlashRound _round, uint256 _epochIndex) public view returns (Epoch epochNumber) {
```
The naming sounds like it was slashing, but not necessarily is the thing to slash, if that makes sense?

spalladino replied:
Renamed to `getSlashTargetEpoch` (thanks Claude for the suggestion)

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+  /**
+   * @notice Get round data for a specific round, loading from circular storage and decompressing it
+   * @param _round The round number to retrieve data for
+   * @param _currentRound The current round number (so we dont try loading data outside the valid roundabout range)
```
If you could a comment to why you are passing the `_currentRound` that would be useful. Assuming that it is because of costs, to avoid re-computing it?

spalladino replied:
Yep. Comment added.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   * @param _round The round number to load committees for
+   * @return committees Array of committees, one for each epoch in the round
+   */
+  function getCommitteesSlashed(SlashRound _round) internal returns (address[][] memory committees) {
```
As the other, the naming is slightly confusing as these guys might not be slashes, but merely the people that are up for "debate".

spalladino replied:
Renamed to `getSlashTargetCommittees`

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+
+    for (uint256 epochIndex = 0; epochIndex < ROUND_SIZE_IN_EPOCHS; epochIndex++) {
+      Epoch epoch = getEpochSlashed(_round, epochIndex);
+      IValidatorSelection rollup = IValidatorSelection(INSTANCE);
```
The `rollup` can be prepared outside of the loop 🤷

spalladino replied:
Done

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+    );
+
+    uint256 voteCount = _roundData.voteCount;
+    SlashRound roundNumber = _roundData.roundNumber;
```
You can delay the `roundNumber` read to after checking the votecount == 0.

spalladino replied:
It's just reading from memory, but fair point

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+          uint256 committeeIndex = i / COMMITTEE_SIZE; // Which epoch's committee
+          uint256 validatorIndex = i % COMMITTEE_SIZE; // Position within committee
+          address validator = _committees[committeeIndex][validatorIndex];
+          uint256 slashAmount = SLASHING_UNIT * j; // Convert units to actual slash amount
```
Should the slashing unit really be a map instead? So the amount is based on `j` and that map? Mainly coming down to, how different is the absolute worst one can do and what is the smallest offsense

spalladino replied:
It's an option, yeah. However, today we only define two amounts for slashing: the entire activation threshold or 1/6th of it. So I figured this was simpler.

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+    }
+
+    // Mark round as executed to prevent re-execution
+    setRoundData(_round, roundData.lastVoteSlot, roundData.voteCount, true);
```
Would move this up above the `.slash` just to avoid any considerations of re-entry that way. Would require a whole slew of things to be broken, but it is one of those that always get pointed at.

spalladino replied:
Good call

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+   *      This is used to verify that vote signatures come from the authorized proposer.
+   * @return The address of the current slot's designated proposer
+   */
+  function getCurrentProposer() internal returns (address) {
```
Naming not following our style of `_` prefix for internal functions for contract functinos.

spalladino replied:
I'm surprised solhint didnt pick it up. Updated.

LHerskind replied:
We went back to a standard version of solhint as there kept being fuckups between what ci ran with and what you got locally, so some of our custom rules are not there atm 😬

LHerskind commented on l1-contracts/src/core/slashing/ConsensusSlashingProposer.sol:
```diff
@@ -0,0 +1,864 @@
+    if (actions.length > 0) {
+      // Deploy payload contract and execute slashes
+      IPayload slashPayload = deploySlashPayload(_round, actions);
+      SLASHER.slash(slashPayload);
```
With tally and all that going on in here, do we have a clear idea about how many of the potential actors we can actually punish? 

Say that a cartel is to be slashed, but some other people did things at the same time such that it is questionable whether they were online or not. It could be in the interest of the cartel to start voting much more aggressively to slash everyone (including themselves) if they can move some of the "maybe" beyond the threshold and have the execution run out of gas.

spalladino replied:
> do we have a clear idea about how many of the potential actors we can actually punish

Afraid not, I need to add more benchmarks to have a good idea of gas expenses.

LHerskind reviewed (commented):
Haven't read through the tests yet, but got some mostly tiny stuff for code.

spalladino commented:
@LHerskind I've addressed all comments and renamed `ConsensusSlash` to `TallySlash` as discussed. As for gas usage, I tweaked `test/governance/scenario/slashing/TallySlashing.t.sol` with:

```
uint256 constant VALIDATOR_COUNT = 128;
uint256 constant COMMITTEE_SIZE = 48;
uint256 constant ROUND_SIZE_IN_EPOCHS = 4;
```

Gas usage for slashing 12 validators is 6M, while gas for slashing everyone on the set (total 192 slashes) was 35M. So we should be able to slash 70-80 validators within the 16M gas limit from [EIP7825](https://eips.ethereum.org/EIPS/eip-7825). This seems expensive but not enough to be a security risk. Let me know your thoughts.

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -166,6 +175,15 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
+
+    if (_slashing == TestSlash.TALLY) {
+      // For tally slashing, we need a round size that's a multiple of epoch duration
+      uint256 tallyRoundSize = EPOCH_DURATION * 2; // 64; // 2 * EPOCH_DURATION (32) = 64
```
I updated this to compute instead of hardcoded.

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -353,41 +371,122 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
   }
 
-  function benchmark(bool _slashing) public {
+  function proposeWithTallyVote(Block memory b, address proposer) internal {
```
Combined the functions down here and moved it into a multicall.

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -353,41 +371,122 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
     // Loop through all of the L1 metadata
     for (uint256 i = 0; i < l1Metadata.length; i++) {
-      if (rollup.getPendingBlockNumber() >= 100) {
+      if (rollup.getPendingBlockNumber() >= 200) {
```
Need more than 100, when we have an epoch size of 48 and also are spending some time on warming up 😬 (note that the 200 here is block and not slots, so together with the `nextSlot` valuese above it is the reason we don't need to go much higher)

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -415,10 +514,23 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
+          SlashRound slashRound = TallySlashingProposer(slashingProposer).getCurrentRound();
+          // We are offset + 1, because the first round after the offset is used entirely on warming the storage up, so
+          // we don't get a off-balance update
+          if (SlashRound.unwrap(slashRound) >= 3) {
```
also updated down here, gotta be after the pre-heating.

LHerskind commented on l1-contracts/test/benchmark/happy.t.sol:
```diff
@@ -353,41 +371,122 @@ contract BenchmarkRollupTest is FeeModelTestPoints, DecoderBase {
+   * @return Encoded vote data
+   */
+  function createTallyVoteData(uint8[] memory slashAmounts) internal pure returns (bytes memory) {
+    require(slashAmounts.length % 2 == 0, "Vote data must have even number of validators");
```
Updated this one such that we write non-zero data

LHerskind commented on l1-contracts/src/core/slashing/TallySlashingProposer.sol:
We should have a function to read the votes 👀 right now it don't seem like we can access the votes without doing direct storage loads.

LHerskind commented on l1-contracts/src/core/slashing/TallySlashingProposer.sol:
```diff
@@ -0,0 +1,872 @@
+
+    return (roundData.executed, isReady, roundData.voteCount);
+  }
+
```
I added this one.

LHerskind commented on l1-contracts/src/core/slashing/TallySlashingProposer.sol:
```diff
@@ -0,0 +1,882 @@
+   * @param _index The index to retrieve votes for
+   * @return The votes retrieved
+   */
+  function getVotes(SlashRound _round, uint256 _index) external view returns (bytes memory) {
```
Added this one to address my own comment.

LHerskind commented on l1-contracts/test/governance/scenario/slashing/TallySlashing.t.sol:
```diff
@@ -0,0 +1,381 @@
+    _setupCommitteeForSlashing();
+    uint256 howManyToSlash = HOW_MANY_SLASHED;
+
+    // We slash a small amount and see that they are all still validating, but less stake
```
Comment seems to be stale from a copy?

LHerskind reviewed (approved):
Made some minor changes to get the benchmarks back running.

LHerskind committed:
feat: Consensus-based slashing proposer contract

Implementation for the consensus-based slashing on L1 as discussed [here](https://linear.app/aztec-labs/document/slashing-discussion-w33-8810b51b11b0#l1-consensus-db4c0919):

> Build a new SlashingProposer version that does not extend from EmpireBase, but follows same logic on slashing rounds, execution delays, and allowing votes from the current L2 proposer only.

> Each vote is now an array of proposed slashes. We define a slashing-unit value, and every slash is a multiple of this unit. Slashes are represented as a bit vector, where the i-th 4-bits (or 2, or 8, depending on how much granularity we want) indicate how many slashing-units proposer i is slashed. An expensive tallying function computes the effective payload by iterating over all votes.

This still keeps the EmpireSlashProposer around. Next step will be to let the `Slasher` choose which implementation to use, so we can keep tests for both, and then rebase this on top of #16284 and tweak the updated slasher client to use this new SlashProposer.

Other changes:
- Updates the `SlashProposer` to be `Ownable` by governance, who can execute slash payloads.
- Modifies the `Slasher` deployment due to contract size limits (the slasher no longer deploys the proposer, instead it needs to be initialized after deployment with the proposer address).
- Renames `ExtRollupLibN` to have more descriptive names.
- Adds env var config for the new proposer flavor.
- Adds type-checking for L1 contract deployer on ts.
